src/openstack_cloud/openstack_cloud.py

```diff
@@ -0,0 +1,597 @@
+
+        with _get_openstack_connection(
+            clouds_config=self._clouds_config, cloud=self._cloud
+        ) as conn:
+            instance_list = self._get_openstack_instances(conn)
+            server_names = set(server.name for server in instance_list)
+
+            server_list = [
+                OpenstackCloud._get_and_ensure_unique_server(conn, name) for name in server_names
+            ]
+            return tuple(
+                OpenstackInstance(server, self.prefix)
+                for server in server_list
+                if server is not None
+            )
+
+    def cleanup(self) -> None:
+        """Cleanup unused key files and openstack keypairs."""
+        with _get_openstack_connection(
+            clouds_config=self._clouds_config, cloud=self._cloud
+        ) as conn:
+            instances = self._get_openstack_instances(conn)
+            exclude_list = [server.name for server in instances]
+            self._cleanup_key_files(exclude_list)
+            self._cleanup_openstack_keypairs(conn, exclude_list)
+
+    def get_server_name(self, instance_id: str) -> str:
+        """Get server name on OpenStack.
+
+        Args:
+            instance_id: ID used to identify a instance.
+
+        Returns:
+            The OpenStack server name.
+        """
+        return f"{self.prefix}-{instance_id}"
+
+    def _cleanup_key_files(self, exclude_instances: Iterable[str]) -> None:
+        """Delete all SSH key files except the specified instances.
+
+        Args:
+            exclude_instances: The keys of these instance will not be deleted.
+        """
+        logger.info("Cleaning up SSH key files")
+        exclude_filename = set(
+            OpenstackCloud._get_key_path(instance) for instance in exclude_instances
+        )
+
+        total = 0
+        deleted = 0
+        for path in _SSH_KEY_PATH.iterdir():
+            # Find key file from this application.
+            if path.is_file() and path.name.startswith(self.prefix) and path.name.endswith(".key"):
+                total += 1
+                if path.name in exclude_filename:
+                    continue
+                path.unlink()
+                deleted += 1
+        logger.info("Found %s key files, clean up %s key files", total, deleted)
+
+    def _cleanup_openstack_keypairs(
+        self, conn: OpenstackConnection, exclude_instances: Iterable[str]
+    ) -> None:
+        """Delete all OpenStack keypairs except the specified instances.
+
+        Args:
+            conn: The Openstack connection instance.
+            exclude_instances: The keys of these instance will not be deleted.
+        """
+        logger.info("Cleaning up openstack keypairs")
+        keypairs = conn.list_keypairs()
+        for key in keypairs:
+            # The `name` attribute is of resource.Body type.
+            if key.name and str(key.name).startswith(self.prefix):
+                if str(key.name) in set(exclude_instances):
```

Comment:
Nit: should the `set(exclude_instances)` be done once at the beginning? The reason is because an iterable can be exhausted after first pass.

Reply:
Good idea.